react_ujs/index.js

```diff
@@ -164,19 +161,44 @@ var ReactRailsUJS = {
     detectEvents(this)
   },
 
-  unmountRoot: function(node) {
-    var targetRoots = this.roots.filter(
+  findOrCreateRoot: function(node) {
+    var root = this.findRoot(node);
+    if (!root) {
+      root = createReactRootLike(node);
+      if(supportsRootApi) {
+        this.roots.push({"node": node, "root": root})
+      }
+    }
+
+    return root;
+  },
+
+  findRoot: function(node) {
+    if (!supportsRootApi) {
+      return;
+    }
+    var rootElement = this.roots.find(
       function(rootElement) {
         return rootElement["node"] && (rootElement["node"] === node)
       }
-    )
-    targetRoots.forEach(
+    );
+    if (rootElement) {
+      return rootElement["root"];
+    }
+  },
+
+  unmountRoot: function(node) {
+    var targetRoot = this.findRoot(node);
+    if (targetRoot) {
+      return;
+    }
+
+    targetRoot["root"].unmount();
+    this.roots = this.roots.filter(
       function(rootElement) {
-        rootElement["root"].unmount();
-        rootElement["root"] = null;
-        rootElement["node"] = null;
+        return rootElement["node"] !== node
       }
-    )
+    );
```

Comment:
Is it possible to have a test that fails without this change? And passes with this change?

Can you extract some code from
> I made a [reproduction repo](https://github.com/diogobeda/react-rails-18-root-bug) to showcase the bug happening and as a test case for the fix. Feel free to try it out for yourselves.

into https://github.com/reactjs/react-rails/blob/master/test/dummy

and add a test?

Reply:
Sounds good! I'll do that in the next few days.

Reply:
Hey @justin808, the dummy project's appraisals are missing some of the dependencies I've used in the reproduction case, such as Turbo and sqlite for creating the Timer records. Can I add those or would you prefer another approach?

Reply:
@diogobeda feel free to add them.